File: Java/lista_de_exercicios_01/lista_de_exercicios_01/src/Classes/Animal.java
package Classes;

public abstract class Animal {
    String especie;

    public Animal(String especie) {
        this.especie = especie;
    }

    public String getEspecie() {
        return especie;
    }

    public void setEspecie(String especie) {
        this.especie = especie;
    }

    public String emitirSom() {
        return " ";
    }

    public String comer() {
        return "Estou comendo!";
    }
}

File: Golang/lista_de_exercicios_01/classes/animal/animal.go
package animal

import "fmt"

type Barulho interface {
	EmitirBarulho()
}

type animal struct {
	especie string
	som     string
}

func (a animal) Comer() {
	fmt.Printf("Estou comendo!\n")
}

func (a animal) EmitirSom() {
	fmt.Printf("%s\n", a.som)
}

type cachorro struct {
	animal
}

func NewCachorro() cachorro {
	c := cachorro{
		animal{
			especie: "Cachorro",
			som:     "Au au!",
		},
	}
	return c
}

func (c cachorro) EmitirBarulho() {
	fmt.Printf("O cachorro faz au au!\n")
}

type gato struct {
	animal
}

func NewGato() gato {
	g := gato{
		animal{
			especie: "Gato",
			som:     "Miau!",
		},
	}
	return g
}

func (g gato) EmitirBarulho() {
	fmt.Printf("O gato faz miau!\n")
}
